crypto/fn/fn_addsub.c

```diff
@@ -33,33 +33,53 @@ int OSSL_FN_add(OSSL_FN *r, const OSSL_FN *a, const OSSL_FN *b)
     size_t max = a->dsize;
     size_t min = b->dsize;
     size_t rs = r->dsize;
+    const OSSL_FN_ULONG *ap = a->d;
+    const OSSL_FN_ULONG *bp = b->d;
+    OSSL_FN_ULONG *rp = r->d;
+
 
     /*
-     * 'r' must be able to contain the result.  This is on the caller.
+     * Three stages, after which there's a possible return.
+     * Each stage is limited by the number of remaining limbs
+     * in |r|.
+     *
+     * For each stage, |stage_limbs| is used to hold the number
+     * of limbs being treated in that stage, and |borrow| is
+     * used to transport the borrow from one stage to the other.
      */
-    if (!ossl_assert(max <= rs)) {
-        ERR_raise(ERR_LIB_OSSL_FN, OSSL_FN_R_RESULT_ARG_TOO_SMALL);
-        return 0;
-    }
-
-    const OSSL_FN_ULONG *ap = a->d;
-    const OSSL_FN_ULONG *bp = b->d;
+    size_t stage_limbs;
+    OSSL_FN_ULONG carry;
 
-    OSSL_FN_ULONG *rp = r->d;
-    OSSL_FN_ULONG carry = bn_add_words(rp, ap, bp, (int)min);
+    /* Stage 1 */
 
+    stage_limbs = (min > rs) ? rs : min;
+    carry = bn_add_words(rp, ap, bp, (int)stage_limbs);
     rp += min;
     ap += min;
+    if (stage_limbs == rs)
+        return 1;
```

Comment:
Actually this could be moved to before the above two increment lines, and then adding `min` seems to make more sense.